module/executiondatasync/tracker/storage.go

```diff
@@ -0,0 +1,454 @@
+		}
+
+		fulfilledHeight = binary.LittleEndian.Uint64(fulfilledHeightValue)
+
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	return fulfilledHeight, nil
+}
+
+func (s *storage) trackBlob(txn *badger.Txn, blockHeight uint64, c cid.Cid) error {
+	if err := txn.Set(makeBlobRecordKey(blockHeight, c), nil); err != nil {
+		return fmt.Errorf("failed to add blob record: %w", err)
+	}
+
+	latestHeightKey := makeLatestHeightKey(c)
+	item, err := txn.Get(latestHeightKey)
+	if err != nil {
+		if !errors.Is(err, badger.ErrKeyNotFound) {
+			return fmt.Errorf("failed to get latest height: %w", err)
+		}
+	} else {
+		value, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve latest height value: %w", err)
+		}
+
+		// don't update the latest height if there is already a higher block height containing this blob
+		latestHeight := binary.LittleEndian.Uint64(value)
+		if latestHeight >= blockHeight {
+			return nil
+		}
+	}
+
+	latestHeightValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(latestHeightValue, blockHeight)
+
+	if err := txn.Set(latestHeightKey, latestHeightValue); err != nil {
+		return fmt.Errorf("failed to set latest height value: %w", err)
+	}
+
+	return nil
+}
+
+func (s *storage) trackBlobs(blockHeight uint64, cids ...cid.Cid) error {
+	cidsPerBatch := 16
+	maxCidsPerBatch := getBatchItemCountLimit(s.db, 2, blobRecordKeyLength+latestHeightKeyLength+8)
+	if maxCidsPerBatch < cidsPerBatch {
+		cidsPerBatch = maxCidsPerBatch
+	}
+
+	for len(cids) > 0 {
+		batchSize := cidsPerBatch
+		if len(cids) < batchSize {
+			batchSize = len(cids)
+		}
+		batch := cids[:batchSize]
+
+		if err := retryOnConflict(s.db, func(txn *badger.Txn) error {
+			for _, c := range batch {
+				if err := s.trackBlob(txn, blockHeight, c); err != nil {
+					return fmt.Errorf("failed to track blob %s: %w", c.String(), err)
+				}
+			}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		cids = cids[batchSize:]
+	}
+
+	return nil
+}
+
+func (s *storage) batchDelete(deleteInfos []*deleteInfo) error {
+	return s.db.Update(func(txn *badger.Txn) error {
+		for _, dInfo := range deleteInfos {
+			if err := txn.Delete(makeBlobRecordKey(dInfo.height, dInfo.cid)); err != nil {
+				return fmt.Errorf("failed to delete blob record for Cid %s: %w", dInfo.cid.String(), err)
+			}
+
+			if dInfo.deleteLatestHeightRecord {
+				if err := txn.Delete(makeLatestHeightKey(dInfo.cid)); err != nil {
+					return fmt.Errorf("failed to delete latest height record for Cid %s: %w", dInfo.cid.String(), err)
+				}
+			}
+		}
+
+		return nil
+	})
+}
+
+func (s *storage) batchDeleteItemLimit() int {
+	itemsPerBatch := 256
+	maxItemsPerBatch := getBatchItemCountLimit(s.db, 2, blobRecordKeyLength+latestHeightKeyLength)
+	if maxItemsPerBatch < itemsPerBatch {
+		itemsPerBatch = maxItemsPerBatch
+	}
+	return itemsPerBatch
+}
+
+func (s *storage) Prune(height uint64) error {
```

Comment:
Please document this function. For example:
* what data is pruned?
* pruning locks the tracker (other writes are blocked during the pruning)
* prunes up to and including `height`
* what happens if we prune a height that isn't fulfilled?